File: plugins/totem-enforce/hooks/totem-enforce.py
#!/usr/bin/env python3
"""Totem enforcement hook. Reads JSON from stdin, checks memory, decides allow/block.

Used by: OpenCode, Claude Code, Kimi Code.
Smart blocking: only block if memory is non-empty AND related to the query.
Detects sub-commands in bash (grep, find, cat, etc.) and searches content, not just cmd: tags.
"""

import json
import os
import re
import subprocess
import sys
import tempfile
from pathlib import Path

# Sub-commands that search/read file content
SUBCMDS = re.compile(
    r"^(grep|find|cat|head|tail|wc|sort|uniq|awk|sed|less|more|diff|comm|xargs|file|rg|ag|ack|jq)$"
)
STOP_WORDS = {"the", "and", "for", "not", "with", "from", "this", "that"}
FTS5_SPECIAL = re.compile(r'[: "+*^()~]')


def get_state_path() -> Path:
    """Get path to state file for this session."""
    session_id = os.environ.get("SESSION_ID", os.getppid())
    return Path(tempfile.gettempdir()) / f"totem-hook-state-{session_id}.json"


def load_state() -> dict:
    """Load hook state from temp file."""
    path = get_state_path()
    if path.exists():
        try:
            return json.loads(path.read_text())
        except (json.JSONDecodeError, OSError):
            pass
    return {"searched_this_turn": {}}


def save_state(state: dict) -> None:
    """Save hook state to temp file."""
    path = get_state_path()
    try:
        path.write_text(json.dumps(state))
    except OSError:
        pass


def check_memory(query: str, project_dir: str, *, use_cmd_tag: bool = True) -> bool:
    """Check if totem has memory related to this query."""
    try:
        cmd = ["totem", "search", "--query", query, "--limit", "1", "--project", project_dir]
        if use_cmd_tag:
            cmd.extend(["--tags", f"cmd:{query}"])
        result = subprocess.run(cmd, capture_output=True, text=True, timeout=5)
        if result.returncode != 0:
            return False
        items = json.loads(result.stdout) if result.stdout.strip() else []
        return len(items) > 0
    except (subprocess.TimeoutExpired, FileNotFoundError, json.JSONDecodeError):
        return False  # Fail open


def tokenize_bash(command: str) -> list[str]:
    """Extract meaningful words from a bash command.

    For sub-commands (grep, find, cat, etc.), returns words from the full command.
    For other commands, returns just the command name.
    """
    cmd = command.strip()
    # Strip leading cd/chains
    cmd = re.sub(r"^cd\s+\S+\s*&&\s*", "", cmd)
    cmd = re.sub(r"^cd\s+\S+\s*;\s*", "", cmd)
    parts = cmd.split()
    if not parts:
        return []
    first = parts[0].split("/")[-1].lower()

    if SUBCMDS.match(first):
        # Sub-command: tokenize the full command
        sanitized = FTS5_SPECIAL.sub(" ", cmd)
        words = [w.lower() for w in sanitized.split() if len(w) > 2 and w.lower() not in STOP_WORDS]
        return list(dict.fromkeys(words))  # dedupe, preserve order
    return [first] if len(first) > 2 else []


def build_search_key(tool_name: str, tool_input: dict) -> str:
    """Build a search key from tool name and input."""
    if tool_name in ("Grep", "grep"):
        return f"grep:{tool_input.get('pattern', tool_input.get('regex', ''))}"
    if tool_name in ("Glob", "glob"):
        return f"glob:{tool_input.get('pattern', '')}"
    if tool_name in ("Read", "read"):
        return f"read:{tool_input.get('filePath', '')}"
    if tool_name in ("Bash", "bash"):
        return f"bash:{tool_input.get('command', '')}"
    return ""


def main() -> None:
    """Main hook entry point."""
    # Read event data from stdin
    input_data = sys.stdin.read()
    try:
        payload = json.loads(input_data)
    except json.JSONDecodeError:
        sys.exit(0)  # Fail open

    tool_name = payload.get("tool_name", "")
    tool_input = payload.get("tool_input", {})
    project_dir = payload.get("cwd", os.getcwd())

    # Only intercept search/read tools
    if tool_name not in ("Grep", "grep", "Glob", "glob", "Read", "read", "Bash", "bash"):
        sys.exit(0)  # Allow non-intercepted tools

    # Build search key
    search_key = build_search_key(tool_name, tool_input)
    if not search_key:
        sys.exit(0)  # Allow if can't build key

    # Load state
    state = load_state()

    # Check if already searched this turn
    if search_key in state.get("searched_this_turn", {}):
        sys.exit(0)  # Allow - already searched

    # Check memory
    if tool_name in ("Bash", "bash"):
        command = tool_input.get("command", "")
        words = tokenize_bash(command)
        has_memory = False
        for w in words:
            # Sub-command args: search general memory. Command names: search cmd: tags
            first_word = command.strip().split()[0].split("/")[-1].lower() if command.strip() else ""
            is_subcmd = bool(SUBCMDS.match(first_word))
            if check_memory(w, project_dir, use_cmd_tag=not is_subcmd):
                has_memory = True
                break
    else:
        has_memory = check_memory(search_key, project_dir)
    if not has_memory:
        sys.exit(0)  # No memory → allow

    # Memory exists → block
    state.setdefault("searched_this_turn", {})[search_key] = "now"
    save_state(state)

    # Determine tool category for redirect message
    if tool_name in ("Grep", "grep", "Glob", "glob"):
        redirect = "memory_search_tool"
    elif tool_name in ("Read", "read"):
        redirect = "engineering_context_tool"
    elif tool_name in ("Bash", "bash"):
        redirect = "memory_commands_tool"
    else:
        redirect = "memory_search_tool"

    reason = (
        f"Totem has memory about this. Use {redirect} first. "
        f"Only {tool_name} the codebase if memory returns nothing relevant."
    )

    # Output JSON decision (works for Claude Code and Kimi Code)
    print(json.dumps({
        "hookSpecificOutput": {
            "hookEventName": "PreToolUse",
            "permissionDecision": "deny",
            "permissionDecisionReason": reason,
        }
    }))
    sys.exit(0)


if __name__ == "__main__":
    main()

File: plugins/totem-enforce/.opencode/plugins/totem-enforce.js
import { execFileSync } from "child_process"

// Totem enforcement plugin for OpenCode.
// 1. Read gate: block read/grep/glob/bash when totem has relevant memory,
//    redirecting to the memory tools. Retry after checking memory is allowed.
// 2. Read commit-gate: after a successful read, all non-totem tools are
//    blocked until totem_register_file_read_tool is called.
// 3. Write commit-gate: after edit/write, blocked until
//    totem_register_file_write_tool is called.
// State is per-session (keyed by sessionID), reset on session.idle.

const sessions = new Map()

const FTS5_SPECIAL = /[:"'+*^()~]/g
const STOP_WORDS = new Set(["the", "and", "for", "not", "with", "from", "this", "that"])
const SUBCMDS = /^(grep|find|cat|head|tail|wc|sort|uniq|awk|sed|less|more|diff|comm|xargs|file|rg|ag|ack|jq)$/

function getState(sessionID) {
  let s = sessions.get(sessionID)
  if (!s) {
    s = { searched: {}, pendingRead: null, pendingWrite: null }
    sessions.set(sessionID, s)
  }
  return s
}

function sanitizeFts5(q) { return q.replace(FTS5_SPECIAL, " ").trim() }

function tokenize(raw) {
  return [...new Set(
    sanitizeFts5(raw).split(/[/\\._\- ='"{},]+/)
      .map(w => w.toLowerCase())
      .filter(w => w.length > 2 && !STOP_WORDS.has(w))
  )]
}

function tokenizeBash(command) {
  let cmd = command.trim()
    .replace(/^cd\s+\S+\s*&&\s*/, "")
    .replace(/^cd\s+\S+\s*;\s*/, "")
  const first = cmd.split(/\s+/)[0]?.split("/").pop()?.toLowerCase() || ""
  if (SUBCMDS.test(first)) return tokenize(cmd)
  return first.length > 2 ? [first] : []
}

function totemSearch(dir, query, { types, tags } = {}) {
  try {
    // Arg array, no shell — agent-controlled queries can't inject.
    // --project is a group-level option: it must precede the subcommand.
    const args = ["--project", dir, "search", "--query", query, "--limit", "1"]
    if (types) args.push("--types", types)
    if (tags) args.push("--tags", tags)
    const r = execFileSync("totem", args, { timeout: 5000, encoding: "utf-8", cwd: dir, stdio: ["pipe", "pipe", "pipe"] })
    return JSON.parse(r || "[]").length > 0
  } catch {
    return false // Fail open
  }
}

function hasMemoryFor(dir, tool, args) {
  if (tool === "read") {
    const fp = args?.filePath || ""
    if (!fp) return false
    const name = fp.split("/").pop()
    return totemSearch(dir, fp, { types: "implementation" })
      || (name && totemSearch(dir, name, { types: "implementation" }))
  }
  if (tool === "grep" || tool === "glob") {
    for (const w of tokenize(args?.pattern || args?.regex || "")) {
      if (totemSearch(dir, w)) return true
    }
    return false
  }
  if (tool === "bash") {
    const command = args?.command || ""
    const first = command.trim().split(/\s+/)[0]?.split("/").pop()?.toLowerCase() || ""
    if (!SUBCMDS.test(first)) {
      return first && totemSearch(dir, first, { tags: `cmd:${first}` })
    }
    for (const w of tokenizeBash(command)) {
      if (totemSearch(dir, w)) return true
    }
    return false
  }
  return false
}

function buildSearchKey(tool, args) {
  if (tool === "grep") return `grep:${args?.pattern || args?.regex || ""}`
  if (tool === "glob") return `glob:${args?.pattern || ""}`
  if (tool === "read") return `read:${args?.filePath || ""}`
  if (tool === "bash") return `bash:${args?.command || ""}`
  return ""
}

export const TotemEnforce = async ({ directory } = {}) => {
  const dir = directory || process.cwd()
  return {
    "tool.execute.before": async (input, output) => {
      const tool = input.tool
      const state = getState(input.sessionID || "default")

      // 1. Register calls clear their gate.
      if (tool === "totem_register_file_read_tool") { state.pendingRead = null; return }
      if (tool === "totem_register_file_write_tool") { state.pendingWrite = null; return }

      // 2. Commit-gate: pending registration blocks all non-totem tools.
      if (!tool.startsWith("totem_")) {
        if (state.pendingRead) {
          throw new Error(
            `You read ${state.pendingRead}. You MUST call register_file_read_tool ` +
            `with what you learned (path, subject, kind, statement, tags) before doing anything else.`
          )
        }
        if (state.pendingWrite) {
          throw new Error(
            `You modified ${state.pendingWrite}. You MUST call register_file_write_tool ` +
            `documenting what changed and why before doing anything else.`
          )
        }
      }

      // Remember read/edit/write target paths so the after-hook can arm the
      // gate on success (the after-hook does not receive args).
      if (["read", "edit", "write"].includes(tool) && output.args?.filePath) {
        state.lastCall = state.lastCall || {}
        state.lastCall[input.callID] = { tool, filePath: output.args.filePath }
      }

      // 3. Memory gates for search/read tools.
      if (!["read", "grep", "glob", "bash"].includes(tool)) return
      const sk = buildSearchKey(tool, output.args || {})
      if (!sk || state.searched[sk]) return
      if (!hasMemoryFor(dir, tool, output.args || {})) return
      state.searched[sk] = true
      const redirect = tool === "read" ? "engineering_context_tool (with paths=[...]) or memory_search_tool"
        : tool === "bash" ? "memory_commands_tool"
        : "memory_search_tool"
      throw new Error(
        `Totem has memory about this. Use ${redirect} first. ` +
        `Only ${tool} the codebase if memory returns nothing relevant. Do not bypass via another tool.`
      )
    },

    "tool.execute.after": async (input, output) => {
      const state = getState(input.sessionID || "default")
      const call = state.lastCall?.[input.callID]
      if (!call) return
      delete state.lastCall[input.callID]
      // Don't arm gates for failed tool calls.
      if (output.error) return
      if (call.tool === "read") state.pendingRead = call.filePath
      if (call.tool === "edit" || call.tool === "write") state.pendingWrite = call.filePath
    },

    event: async ({ event }) => {
      // Clear only the session that went idle, not all sessions.
      if (event.type === "session.idle") {
        const sid = event.properties?.sessionID
        if (sid) sessions.delete(sid)
        else sessions.clear()
      }
    },
  }
}

export default TotemEnforce
